File: packages/textlayout/src/textLineBreaks.ts
/**
 * Binary search for the first break index at or after `startIndex`.
 * Returns -1 when no such break exists. The `lineBreaks` array must be sorted
 * in ascending order (as produced by `getTextLineBreaks`).
 */
export function getTextLineBreakIndex(lineBreaks: readonly number[], startIndex = 0): number {
  if (lineBreaks.length === 0) return -1;
  let lo = 0;
  let hi = lineBreaks.length - 1;
  let result = -1;
  while (lo <= hi) {
    const mid = (lo + hi) >>> 1;
    if (lineBreaks[mid] >= startIndex) {
      result = lineBreaks[mid];
      hi = mid - 1;
    } else {
      lo = mid + 1;
    }
  }
  return result;
}

export function getTextLineBreaks(out: number[], text: string): void {
  out.length = 0;
  let index = -1;

  while (index < text.length) {
    const lf = text.indexOf('\n', index + 1);
    const cr = text.indexOf('\r', index + 1);

    if (lf === -1 && cr === -1) break;

    index = cr === -1 ? lf : lf === -1 ? cr : Math.min(cr, lf);
    out.push(index);
  }
}

File: crates/flighthq-text-layout/src/text_line_breaks.rs
/// Returns the first line-break index in `line_breaks` that is greater than
/// or equal to `start_index`, or `-1` when none exists.
pub fn get_text_line_break_index(line_breaks: &[usize], start_index: usize) -> i64 {
    for &lb in line_breaks {
        if lb >= start_index {
            return lb as i64;
        }
    }
    -1
}

/// Fills `out` with the byte offsets of every `\n` and `\r` character in
/// `text`, in ascending order.
pub fn get_text_line_breaks(out: &mut Vec<usize>, text: &str) {
    out.clear();
    for (i, ch) in text.char_indices() {
        if ch == '\n' || ch == '\r' {
            out.push(i);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn get_text_line_break_index_empty() {
        assert_eq!(get_text_line_break_index(&[], 0), -1);
    }

    #[test]
    fn get_text_line_break_index_found() {
        assert_eq!(get_text_line_break_index(&[3, 7, 12], 5), 7);
    }

    #[test]
    fn get_text_line_break_index_exact() {
        assert_eq!(get_text_line_break_index(&[3, 7], 3), 3);
    }

    #[test]
    fn get_text_line_break_index_none_in_range() {
        assert_eq!(get_text_line_break_index(&[1, 2], 5), -1);
    }

    #[test]
    fn get_text_line_breaks_lf() {
        let mut out = Vec::new();
        get_text_line_breaks(&mut out, "ab\ncd\nef");
        assert_eq!(out, vec![2, 5]);
    }

    #[test]
    fn get_text_line_breaks_cr() {
        let mut out = Vec::new();
        get_text_line_breaks(&mut out, "a\rb\rc");
        assert_eq!(out, vec![1, 3]);
    }

    #[test]
    fn get_text_line_breaks_no_breaks() {
        let mut out = Vec::new();
        get_text_line_breaks(&mut out, "hello");
        assert!(out.is_empty());
    }

    #[test]
    fn get_text_line_breaks_empty_string() {
        let mut out = Vec::new();
        get_text_line_breaks(&mut out, "");
        assert!(out.is_empty());
    }
}
